*: refactor dynamic tso switch

rleungx opened the pull request:
<!--

Thank you for working on PD! Please read PD's [CONTRIBUTING](https://github.com/tikv/pd/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?

<!--

Please create an issue first to describe the problem.
There MUST be one line starting with "Issue Number:  " and 
linking the relevant issues via the "close" or "ref".
For more info, check https://github.com/tikv/pd/blob/master/CONTRIBUTING.md#linking-issues.

-->
Issue Number: ref #8477.

### What is changed and how does it work?

This PR made the following changes:

1. Merge `api` subcommand logic to pd. We keep the subcommand for compatibility but won't use it anymore.
2. Only serverless use env to enable microservice, so I will use a config `enable-multi-timelines` to replace it in the next PR. 
3. Add more details for different scenes in both serverless and non-serverless environments.
4. Whether enabling microservice is not relying on subcommand or system variables. We use `enable-tso-dynamic-switch` for the non-serverless env and `enable-multi-timelines` for the serverless env to manage it.
5. Change the `checkTSOService` and `IsServiceIndependent` logic that needs attention when reviewing.
6. Enable keyspace group by default, which adds group ID and user kind when creating keyspace. It's not ideal but considering the config is not used except for displaying, I made the change.
7. Adapt the above changes for testing.

<!--

You could use the "commit message" code block to add more description to the final commit message.
For more info, check https://github.com/tikv/pd/blob/master/CONTRIBUTING.md#format-of-the-commit-message.

-->

```commit-message
```

### Check List

<!-- Remove the items that are not applicable. -->

Tests

<!-- At least one of these tests must be included. -->

- Unit test

### Release note

<!--

A bugfix or a new feature needs a release note. If there is no need to give a release note, just leave it with the `None`.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

-->

```release-note
None.
```


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Improved TSO provider switching between PD and TSO services based on service readiness.
  * Added safer fallback to PD-backed TSO when forwarding fails or a service becomes unavailable.
  * Added protection against conflicting TSO primaries during timestamp updates.
  * Improved client retries when service modes change.

* **Bug Fixes**
  * Prevented shutdown errors when TSO components are not fully initialized.
  * Embedded member-response errors are now reported correctly.
  * Improved TSO forwarding and failover behavior under concurrent load.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rleungx committed:
remove api mode

Signed-off-by: Ryan Leung <rleungx@gmail.com>

okJiang reviewed (commented):
Is there any test results?

okJiang reviewed (commented):
Please add some description about these changes.

lhy1024 commented on server/cluster/cluster.go:
```diff
@@ -404,53 +404,57 @@ func (c *RaftCluster) Start(s Server, bootstrap bool) (err error) {
-		}
-	} else {
+	servers, err := discovery.Discover(c.etcdClient, constant.SchedulingServiceName)
+	if c.opt.GetMicroServiceConfig().IsSchedulingFallbackEnabled() && (err != nil || len(servers) == 0) {
```
If `SchedulingFallbackEnabled` is false and there is no scheduling server, will we not try to start scheduling service in pd server?

rleungx commented on server/apiv2/handlers/micro_service.go:
```diff
@@ -39,11 +39,6 @@ func RegisterMicroService(r *gin.RouterGroup) {
 // @Router   /ms/members/{service} [get]
 func GetMembers(c *gin.Context) {
 	svr := c.MustGet(middlewares.ServerContextKey).(*server.Server)
-	if !svr.IsPDServiceMode() {
```
After this PR, there won't be pd mode anymore.

rleungx committed:
refactor

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
refactor

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
fix and make comment clearer

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
fix

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
move comments

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
fix

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
remove test

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
tso: fence embedded tso during dynamic switch

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
Merge upstream master into remove-api-mode

rleungx committed:
tso: gate provider changes during dynamic switch

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
tso: fix dynamic switch races

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
tso: probe and retry dynamic provider handoff

Signed-off-by: Ryan Leung <rleungx@gmail.com>

rleungx committed:
tso: avoid unsafe fallback during provider handoff

Keep the current TSO provider unless service discovery confirms that no TSO server is registered. Classify timestamp fence conflicts so embedded PD TSO yields without resigning PD leadership.

Signed-off-by: Ryan Leung <rleungx@gmail.com>
